### DISCUSSION TITLE
Need help with building the code (compiling)

### DISCUSSION THREAD

**Discussion by Olgidos** [rbaron/b-parasite#186] [Q&A]
I currently have PCBs, programmer and everything available. However I am kinda stuck with compiling the firmware. 

My steps:
1. Installed nRF Connect for VS Code v2024.3.25 for VCode
2. Downloaded the b-parasite Git Repo 
3. Clicked on ble.code-workspace -> it opened the workspace including BLE and prstlib. However I dont see a way to build it from here. 
4. I go to the nrf interface, click open existing application 
5. I add a build configuration (see image)
6. Click "Build configuration" -> i get the attached error

From my understanding the open workspace has nothing to do with the application opend with the nrf pluging? 
If so, then don't undertand how to add the prstlib to the nrf application. 

<img width="1503" alt="Screenshot 2024-04-10 at 10 14 40" src="https://github.com/rbaron/b-parasite/assets/25614324/a848c91c-626c-46c0-bb54-43cef453d83a">

Building ble
west build --build-dir /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble --pristine --board nrf52840dk_nrf52840 --no-sysbuild -- -DNCS_TOOLCHAIN_VERSION=NONE -DBOARD_ROOT=/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib

-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble
-- CMake version: 3.21.0
-- Found Python3: /opt/nordic/ncs/toolchains/580e4ef81c/opt/python@3.9/bin/python3.9 (found suitable version "3.9.6", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /Users/olgidos/Library/Caches/zephyr
-- Zephyr version: 3.5.99 (/opt/nordic/ncs/v2.6.0/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: nrf52840dk_nrf52840
-- Found host-tools: zephyr 0.16.5 (/opt/nordic/ncs/toolchains/580e4ef81c/opt/zephyr-sdk)
-- Found toolchain: zephyr 0.16.5 (/opt/nordic/ncs/toolchains/580e4ef81c/opt/zephyr-sdk)
-- Found Dtc: /opt/nordic/ncs/toolchains/580e4ef81c/bin/dtc (found suitable version "1.6.1", minimum required is "1.4.6") 
-- Found BOARD.dts: /opt/nordic/ncs/v2.6.0/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Found devicetree overlay: /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/nrf52840dk_nrf52840.overlay
-- Generated zephyr.dts: /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/dts.cmake

warning: UART_CONSOLE (defined at drivers/console/Kconfig:43) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: SERIAL (=n), SERIAL_HAS_DRIVER (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_UART_CONSOLE and/or look up UART_CONSOLE in
the menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration
Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful too.

Parsing /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/Kconfig
Loaded configuration '/opt/nordic/ncs/v2.6.0/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration '/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/prj.conf'
Configuration saved to '/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/.config'
Kconfig header saved to '/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /opt/nordic/ncs/toolchains/580e4ef81c/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/nordic/ncs/toolchains/580e4ef81c/opt/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Using ccache: /opt/nordic/ncs/toolchains/580e4ef81c/bin/ccache
CMake Warning at /opt/nordic/ncs/v2.6.0/zephyr/CMakeLists.txt:862 (message):
  No SOURCES given to Zephyr library: drivers__console

  Excluding target from build.

CMake Warning at /opt/nordic/ncs/v2.6.0/zephyr/CMakeLists.txt:1957 (message):
  __ASSERT() statements are globally ENABLED

Loading Zephyr default modules (Zephyr base (cached)).
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build
-- west build: building application
[1/206] Preparing syscall dependency handling

[4/206] Generating include/generated/version.h
-- Zephyr version: 3.5.99 (/opt/nordic/ncs/v2.6.0/zephyr), build: v3.5.99-ncs1
[198/206] Building C object prstlib/CMakeFiles/prstlib.dir/src/adc.c.obj
FAILED: prstlib/CMakeFiles/prstlib.dir/src/adc.c.obj 
ccache /opt/nordic/ncs/toolchains/580e4ef81c/opt/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DNRF52840_XXAA -D_ANSI_SOURCE -D_FORTIFY_SOURCE=1 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/src -I/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/include -I/opt/nordic/ncs/v2.6.0/nrf/drivers/mpsl/clock_control -I/opt/nordic/ncs/v2.6.0/zephyr/include -I/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/include/generated -I/opt/nordic/ncs/v2.6.0/zephyr/soc/arm/nordic_nrf/nrf52 -I/opt/nordic/ncs/v2.6.0/zephyr/lib/libc/newlib/include -I/opt/nordic/ncs/v2.6.0/zephyr/soc/common/nordic_nrf/. -I/opt/nordic/ncs/v2.6.0/zephyr/soc/arm/nordic_nrf/common/. -I/opt/nordic/ncs/v2.6.0/zephyr/subsys/bluetooth -I/opt/nordic/ncs/v2.6.0/nrf/include -I/opt/nordic/ncs/v2.6.0/nrf/lib/multithreading_lock/. -I/opt/nordic/ncs/v2.6.0/nrf/subsys/bluetooth/controller/. -I/opt/nordic/ncs/v2.6.0/nrf/tests/include -I/opt/nordic/ncs/v2.6.0/modules/hal/cmsis/CMSIS/Core/Include -I/opt/nordic/ncs/v2.6.0/zephyr/modules/cmsis/. -I/opt/nordic/ncs/v2.6.0/modules/hal/nordic/nrfx -I/opt/nordic/ncs/v2.6.0/modules/hal/nordic/nrfx/drivers/include -I/opt/nordic/ncs/v2.6.0/modules/hal/nordic/nrfx/mdk -I/opt/nordic/ncs/v2.6.0/zephyr/modules/hal_nordic/nrfx/. -I/opt/nordic/ncs/v2.6.0/modules/debug/segger/SEGGER -I/opt/nordic/ncs/v2.6.0/modules/debug/segger/Config -I/opt/nordic/ncs/v2.6.0/modules/crypto/tinycrypt/lib/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/fem/common/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/fem/nrf21540_gpio/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/fem/nrf21540_gpio_spi/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/fem/simple_gpio/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/fem/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/fem/include/protocol -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/include -I/opt/nordic/ncs/v2.6.0/nrfxlib/mpsl/include/protocol -I/opt/nordic/ncs/v2.6.0/nrfxlib/softdevice_controller/include -isystem /opt/nordic/ncs/v2.6.0/zephyr/lib/libc/common/include -isystem /opt/nordic/ncs/v2.6.0/nrfxlib/crypto/nrf_cc310_platform/include -fno-strict-aliasing -Os -imacros /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build/zephyr/include/generated/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/opt/nordic/ncs/toolchains/580e4ef81c/opt/zephyr-sdk/arm-zephyr-eabi/arm-zephyr-eabi -imacros /opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble=CMAKE_SOURCE_DIR -fmacro-prefix-map=/opt/nordic/ncs/v2.6.0/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/opt/nordic/ncs/v2.6.0=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=nano.specs -std=c99 -MD -MT prstlib/CMakeFiles/prstlib.dir/src/adc.c.obj -MF prstlib/CMakeFiles/prstlib.dir/src/adc.c.obj.d -o prstlib/CMakeFiles/prstlib.dir/src/adc.c.obj -c /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c
In file included from /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:4:
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:198:36: error: 'DT_N_NODELABEL_soil_calibration_coeffs_P_dry' undeclared here (not in a function)
  198 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                                    ^~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:4343:29: note: in definition of macro 'DT_CAT3'
 4343 | #define DT_CAT3(a1, a2, a3) a1 ## a2 ## a3
      |                             ^~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:21:34: note: in expansion of macro 'DT_PROP'
   21 | static const int dry_coeffs[3] = DT_PROP(DT_NODELABEL(soil_calibration_coeffs), dry);
      |                                  ^~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:198:29: note: in expansion of macro 'DT_CAT'
  198 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:21:42: note: in expansion of macro 'DT_NODELABEL'
   21 | static const int dry_coeffs[3] = DT_PROP(DT_NODELABEL(soil_calibration_coeffs), dry);
      |                                          ^~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:198:36: error: 'DT_N_NODELABEL_soil_calibration_coeffs_P_wet' undeclared here (not in a function)
  198 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                                    ^~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:4343:29: note: in definition of macro 'DT_CAT3'
 4343 | #define DT_CAT3(a1, a2, a3) a1 ## a2 ## a3
      |                             ^~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:22:34: note: in expansion of macro 'DT_PROP'
   22 | static const int wet_coeffs[3] = DT_PROP(DT_NODELABEL(soil_calibration_coeffs), wet);
      |                                  ^~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:198:29: note: in expansion of macro 'DT_CAT'
  198 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:22:42: note: in expansion of macro 'DT_NODELABEL'
   22 | static const int wet_coeffs[3] = DT_PROP(DT_NODELABEL(soil_calibration_coeffs), wet);
      |                                          ^~~~~~~~~~~~
In file included from /opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/toolchain/gcc.h:98,
                 from /opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/toolchain.h:50,
                 from /opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/sys/util.h:18,
                 from /opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:26:
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/device.h:89:41: error: '__device_dts_ord_DT_N_NODELABEL_fast_disch_P_gpios_IDX_0_PH_ORD' undeclared here (not in a function)
   89 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/toolchain/common.h:137:26: note: in definition of macro '_DO_CONCAT'
  137 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/device.h:89:33: note: in expansion of macro '_CONCAT'
   89 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/device.h:216:37: note: in expansion of macro 'DEVICE_NAME_GET'
  216 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/device.h:233:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  233 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/drivers/gpio.h:331:25: note: in expansion of macro 'DEVICE_DT_GET'
  331 |                 .port = DEVICE_DT_GET(DT_GPIO_CTLR_BY_IDX(node_id, prop, idx)),\
      |                         ^~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/drivers/gpio.h:367:9: note: in expansion of macro 'GPIO_DT_SPEC_GET_BY_IDX'
  367 |         GPIO_DT_SPEC_GET_BY_IDX(node_id, prop, 0)
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:25:5: note: in expansion of macro 'GPIO_DT_SPEC_GET'
   25 |     GPIO_DT_SPEC_GET(DT_NODELABEL(fast_disch), gpios);
      |     ^~~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:198:36: error: 'DT_N_NODELABEL_fast_disch_P_gpios_IDX_0_VAL_pin' undeclared here (not in a function)
  198 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                                    ^~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:4352:9: note: in definition of macro 'DT_CAT7'
 4352 |         a1 ## a2 ## a3 ## a4 ## a5 ## a6 ## a7
      |         ^~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree/gpio.h:164:9: note: in expansion of macro 'DT_PHA_BY_IDX'
  164 |         DT_PHA_BY_IDX(node_id, gpio_pha, idx, pin)
      |         ^~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/drivers/gpio.h:332:24: note: in expansion of macro 'DT_GPIO_PIN_BY_IDX'
  332 |                 .pin = DT_GPIO_PIN_BY_IDX(node_id, prop, idx),                 \
      |                        ^~~~~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/drivers/gpio.h:367:9: note: in expansion of macro 'GPIO_DT_SPEC_GET_BY_IDX'
  367 |         GPIO_DT_SPEC_GET_BY_IDX(node_id, prop, 0)
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:25:5: note: in expansion of macro 'GPIO_DT_SPEC_GET'
   25 |     GPIO_DT_SPEC_GET(DT_NODELABEL(fast_disch), gpios);
      |     ^~~~~~~~~~~~~~~~
/opt/nordic/ncs/v2.6.0/zephyr/include/zephyr/devicetree.h:198:29: note: in expansion of macro 'DT_CAT'
  198 | #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
      |                             ^~~~~~
/Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/prstlib/src/adc.c:25:22: note: in expansion of macro 'DT_NODELABEL'
   25 |     GPIO_DT_SPEC_GET(DT_NODELABEL(fast_disch), gpios);
      |                      ^~~~~~~~~~~~
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /opt/nordic/ncs/toolchains/580e4ef81c/bin/cmake --build /Users/olgidos/Documents/GitHub/b-parasite/code/nrf-connect/samples/ble/build

 *  The terminal process terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it.

**Comment by rbaron** [comment]
Hi @Olgidos,

From your description I see two issues:

> 3. Clicked on ble.code-workspace -> it opened the workspace including BLE and prstlib. However I dont see a way to build it from here.

You were close. Once the workspace opened, you had to tap "Add build configuration":

<img width="1503" alt="ss1" src="https://github.com/rbaron/b-parasite/assets/1573409/88d38153-7100-4559-aabf-9145ec6cb73a">

You also need to select the correct board and revision. You were currently building for the `nrf5284dk_nrf52840` (a nordic development, and not a b-parasite board), which is why the build failed. You need to tap "Custom boards", at which point a drop-down will let you select your `b_parasite` board. You also need to punch in the "Revision" (e.g.: `2.0.0`, or `1.2.0` etc). It should look like [this screenshot](https://github.com/rbaron/b-parasite/wiki/How-to-Build-the-Firmware-Samples#with-nrf-connect-for-desktop--visual-studio-code) from the docs (adjusted for your board version & revision).
